### PR TITLE
12

### LINKED ISSUE #12: Feature: User can like or dislike products
**Given** the user is authenticated
**When** the client performs a POST operation to the `/products/n/like` URL
**Then** that preference should be created for the current user

**Given** the user is authenticated
**When** the client performs a DELETE operation to the `/products/n/like` URL
**Then** that preference should be removed for the current user

**Given** the user is authenticated
**When** the client performs a GET operation to the `/products/liked` URL
**Then** the response should contain a list of products liked by the user

### PR DESCRIPTION
**Pull request by devinmgarcia**
Description of PR that completes issue here...

## Changes

- Added 2 custom actions in views/product.py starting at like 307
- Added serializer on same page at line 26
- Added new Model CustomerProductLike

## Requests / Responses

**Request**

POST http://localhost:8000/products/88/like

**Response**

HTTP/1.1 201 CREATED

**Request**

DELETE http://localhost:8000/products/88/like

**Response**

HTTP/1.1 204 NO_CONTENT

**Request**

GET http://localhost:8000/products/liked

**Response**

```json
    {
        "id": 7,
        "customer_id": 7,
        "product": {
            "id": 85,
            "name": "S40",
            "price": 699.63,
            "number_sold": 2,
            "description": "2001 Volvo",
            "quantity": 4,
            "created_date": "2018-12-23",
            "location": "Jayapura",
            "image_path": null,
            "average_rating": "Product has no ratings"
        }
    },
    {
        "id": 8,
        "customer_id": 7,
        "product": {
            "id": 86,
            "name": "1500 Club Coupe",
            "price": 676.11,
            "number_sold": 2,
            "description": "1997 GMC",
            "quantity": 2,
            "created_date": "2019-07-26",
            "location": "Santo Tomas",
            "image_path": null,
            "average_rating": "Product has no ratings"
        }
    },
    {
        "id": 9,
        "customer_id": 7,
        "product": {
            "id": 87,
            "name": "Legend",
            "price": 1019.99,
            "number_sold": 1,
            "description": "1995 Acura",
            "quantity": 1,
            "created_date": "2019-05-01",
            "location": "Titab",
            "image_path": null,
            "average_rating": "Product has no ratings"
        }
    }
```

## Testing

Description of how to test code...

- [ ] Run migrations
- [ ] In postman, run a POST to a few different products to like them:
```
http://localhost:8000/products/86/like
```
http://localhost:8000/products/87/like
```
http://localhost:8000/products/88/like
```
- [ ] Try and like a product that is already liked
- [ ] In postman, run a DELETE to one or more product likes you just created
- [ ] Run a GET to:
```
http://localhost:8000/products/liked
```
- [ ] Ensure that the liked products all get returned with the product info included in the response

## Related Issues

- Fixes #12

### CODE REVIEW

File: bangazonapi/views/product.py [hunk 1]
```diff
@@ -293,3 +303,56 @@ def recommend(self, request, pk=None):
...
+                    return Response(None, status=status.HTTP_201_CREATED)
+                else:
+                    return Response({"message":"Product already liked"}, status=status.HTTP_204_NO_CONTENT)
+            except Exception as ex:
+                return Response({'message': ex.args[0]})
+        elif request.method == "DELETE":
+            try:
+                customer = Customer.objects.get(user=request.auth.user)
+                all_customer_likes = CustomerProductLike.objects.filter(customer_id=customer.id)
+                like = all_customer_likes.get(product_id=pk)
```
Comment (hannahhall) [inline bangazonapi/views/product.py]: The `.get` function can take more than one parameter if you have multiple fields you want to get by. So line 333 could be combined with 334.
If the customer mistakenly likes a product more than once though, the get will probably fail with a `MultipleObjectsReturned` error. To solve that take a look at the django `unique_together` constraint for models.

**Review by hannahhall — CHANGES_REQUESTED** [review]

File: bangazonapi/views/product.py (hunk 1 above)
Comment (devinmgarcia) [reply bangazonapi/views/product.py]: I have combined the get into one line. I already have some code in place to make sure a user cannot like a product more than once. Would I still need the unique together constraint added?

Comment (hannahhall) [reply bangazonapi/views/product.py]: I see now. Sorry I missed that part

**Review by hannahhall — APPROVED** [review]
👍

**Pull request merged**